torch/fx/OVERVIEW.md

```diff
@@ -0,0 +1,121 @@
+
+In this example pipeline, we have a Quantize transformation, which is then composed with a Split transformation, then a Lower to Accelerator transformation. Finally, the transformed Modules are compiled with TorchScript for deployment. This last point emphasizes that not only should FX transforms be composable with each other, but their products are composable with other systems like TorchScript compilation or tracing.
+
+# Symbolic Tracing
+
+## About ##
+
+FX’s front-end makes use of the dynamic nature of Python to intercept call-sites for various entities (PyTorch operators, Module invocations, and Tensor method invocations). This functionality is exposed through an API called `torch.fx.symbolic_trace`.  We can see how this works by way of an example:
+
+```python
+import torch
+
+class MyModule(torch.nn.Module):
+  def __init__(self):
+    super().__init__()
+    self.param = torch.nn.Parameter(
+        torch.rand(3, 4))
+    self.linear = torch.nn.Linear(4, 5)
+
+  def forward(self, x):
+    return self.linear(x + self.param).clamp(min=0.0, max=1.0)
+
+from torch.fx import symbolic_trace
+module = MyModule()
+symbolic_traced : torch.fx.GraphModule = symbolic_trace(module)
+
+input = torch.rand(3, 4)
+torch.testing.assert_allclose(symbolic_traced(input), module(input))
+```
+
+Here, we set up a simple Module that exercises different language features: fetching a parameter, applying an arithmetic operator, applying a submodule (linear), and applying a Tensor method. `symbolic_trace` returns an instance of GraphModule, which is in itself a subclass of `nn.Module`. We can see that the `symbolic_traced` instance runs and returns the same result as the original module instance module.
+
+## Tracer ##
+
+`Tracer` is the class that implements the symbolic tracing functionality of `torch.fx.symbolic_trace`. A call to `symbolic_trace(m)` is equivalent to `Tracer().trace(m)`. Tracer can be subclassed to override various behaviors of the tracing process. The different behaviors that can be overridden are described in the docstrings of the methods on the class.
+
+In the default implementation of `Tracer().trace`, the tracer first creates Proxy objects for all arguments in the `forward` function. (This happens in the call to `create_args_for_root`.) Next, the `forward` function is called with the new Proxy arguments. As the Proxies flow through the program, they record all the operations (`torch` function calls, method calls, and operators) that they touch into the growing FX Graph as Nodes.
+
+## Proxy ##
+
+Proxy objects are Node wrappers used by the Tracer to record operations seen during symbolic tracing. If you're doing graph transforms, you can wrap your own Proxy method around a raw Node so that you can use the overloaded operators to add additional things to a Graph.
```

Comment:
I think the `If you're doing graph transforms...` part makes sense at the end after we've described what Proxies are. It would set it up like "1) here's what these are, 2) here's what you can do with them"

I think it might also make sense to have a small example of using Proxies to, e.g. append additional nodes to a Graph

Reply:
Is this enough of a small example that it should be in torch/fx/examples?